### DISCUSSION TITLE
TypeScript support for send's string type

### DISCUSSION THREAD

**Discussion by mellson** [statelyai/xstate#3212] [Ideas] [answered]
Sometimes I find myself using the short-form send. And while this reads nice, it can catch me out if I forget some data in the payload. And after seeing [this fantastic playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAsiCiA3CA7YUC8AoKUA+UA3jrqaJAFxQBEAYgPL3UDcJuAviQcaWeBFWoAhAIIAlFm1JgAhiAA2AexkATKj16kAxooCuaKil0BbAEYQATq01R210naxYVELfJkXoAM31bgAS0UUKABnVBUAHgAVKP4oCAAPYHCQ2ARkNABtanIIagBdAD4AChIAOgqPAHMQqngkixk-CLgkVGAAGiIoXKoYuPZC+KSU7tkFZTUofxQvSygogAU5JVVbKQB+KEzehdjILvHVqaWVyfypKh3+Pv2IC4BKKkRFfxVWLFy0+mAAC0s2mhMFwiFJcLs6IxJLxOLhuGCejcaKIJPhkSIAFrQzRHSbqBG4HT6YCGEzmKwIxwwj4uNweby+AJBKCKP6WADK4Wid2GyRQKlScB+-wsgOA2VyBRK5UqFhqdQaTWALRAwoBGU63V2-UgtiGiT5ArGZ1UVBmcwsC2WEzWsNIW2ulD2-EOJpO1uOF14V21d0ez1e7ycrJFnP5xWEmOoXUIUCJBhotBESaT1FsDw+AHpM4jdfVgI0-GLomLeaNY7sQgWZlV2F0ddADaMomKJfwpZgoPnC8qWxqY7mBM7dYNPnFu0riyQ++0y-zUhWkVWLDXbB1pzym-OFq2cu2LkMMDuNXOjTIUCASFsG6fUjOsnvIAVNsf2pcoCgIMgLCRDF-LFmOYgHocbnqEADu-jAFovw9IoPT-KE-jGGA8jQBOfhQOBjRgJAloyKYijIHBoQQNAvyKOBCH+KknghIo8jIKkwDwQABp+36sWOuoiJ2GHAMWQpsqK-bolihSsNmpAAHobE4tLuJ4UA+CgfiBMEOgWJ4fhhpEN5bkaQkirukpFKUuAVGU1S1F2ipFhqKpqiJ7T1nc+ojNusa4qa0yzPMpw2io6y8A6voulA3num6XqkD6SINv6UAvG8HyadpwC6RGohYgO8YkomyaFWm7AZlgQA) by @Andarist , I tried playing around with some helpers to alleviate this problem.

Let's play with these events:
```ts
type MyEvents =
  | { type: "Non-optional"; date: Date }
  | { type: "Optional"; date?: Date }
  | { type: "Mixed"; date?: Date; age: number }
  | { type: "Only the type" };
```

If I send the "Non-optional" or "Mixed" events without the payload to my machine, there is a high chance of failure.

So let's try to add some code that could help here:
```ts
// Reusing the ExtractEvent from the linked playground above
type ExtractEvent<
  TEvent extends { type: string },
  TType extends TEvent["type"]
> = TEvent extends any
  ? TType extends TEvent["type"]
    ? TEvent
    : never
  : never;

// This will return a type with only the optional properties
type OptionalKeys<TEvent extends { type: string }> = {
  [Property in keyof TEvent]-?: undefined extends TEvent[Property]
    ? Property
    : never;
}[keyof TEvent];

// Naive implementation of an equals type
type Equals<X, Y> = [X, Y] extends [Y, X] ? true : false;

// This will return falsy if we provide an event with required properties besided the type
type OnlyOptional<TEvent extends { type: string }> = Equals<
  Pick<TEvent, "type">,
  Omit<TEvent, OptionalKeys<TEvent>>
>;

// This is an adjusted version of a send function from the linked playground
declare function send<TType extends MyEvents["type"]>(
  ...args: OnlyOptional<ExtractEvent<MyEvents, TType>> extends true
    ? [type: TType | ExtractEvent<MyEvents, TType>]
    : [type: ExtractEvent<MyEvents, TType>]
): void;

// Now we can use the send function and get some idea if it's safe to do so
send("Optional"); // OK
send("Only the type"); // OK
send("Mixed"); // Errors since there's non optional properties
send("Non-optional"); // Errors since there's non optional properties
send({ type:"Non-optional", date:new Date() }); // OK
```

This idea is very rough, and I would like to get an idea if the team thinks this could be a viable direction. If you think so, I would happily give it some more work and open a PR with a draft 😊 

You can find the code from the idea in [a playground here](https://www.typescriptlang.org/play?#code/PTAEAUBsEME8HMBOB7ArgOwCagGbMaAOICWALgBKoBGomxAzgMar33HLoBcoAFqaQAd6nEPDI9qAOkbIAtsHqlopAKaRY0YsAAei5SuB0mLNh3rAAzACYAjFYBQ9kKABKKlsXTxQpHitAAotqkiNCMpAEAbiropLgosj48DKACMAgoGJjcfILCIADuRZKksAIqTIjEAqQwXpL48MBpcAD8rJAAvDYAZKyM3T1p6J1WQwM2AMQymAYAwgAsACoFACLwAvAAQgCCbHPEOxZzOwDsAJoAqnMAHDvIANKXOwDUzxYAVogAjtAAUjsAGLaACKWwCO3O8HAABkLJdVow-qgdpFiCDGNAAOoqLaArHIHY8HYASRJO0ggNkNj+8CJhx2Vh2Km0AGsdksHh8bAA2dnaZDIOaoaAPYiQAAMWyoOwC5xBHO+EpsLlsEvQ0G05wAagEYTg-qyABqEgCyFhsVDpAGUJZdLg8dlt0Nqts9yHTtQ9wAshTsduBzuQ5iCmS5WQA5J1KdBk6DbTA7VYLHZzHjkgDy8G+4B25FNIO+XlZxDZAFo5jD4KztYR-SQScgjQ9kOdUOAAFpzF4Pa2XU6kVmAjtLS7IHCgyLWhDIOmXACsf212qoDzBPYCAC8VDxMB8STDIuRtb8sdrYDgyg8eBYXuBIlYSYstg8HqyXICJ9qsTDAbAJS8sj+nMEr0FYsg8KagIFMQgJ-IwML0DyJLaH8MLaBmMISIQcwqB8mIPCSfzgDwrLWuSgKICCm4uOQ1pUN8WyEAs5BLDy5xzIQW6yA8BQoUUHwLKgEYFCoVCwC8Ox-C+czejyWJJg8CxYsQmBWIQsAuCCDyQNajBbC8ZYFPQ8AZoy55bC4Dzzpc3yAjw3yYCC0BLApMIuDsBTIExlw8hY2pWGWlzwKQxAkEs847IghAdjYlyurIjDGXMVhYluWL0BGJJYlYqCQKsgUBBmfwCBmNioBmgI2AsAU2JgEoqJuNgqKcsCMIgfyAug4DztoaWyBmTaXCS1zSTwyCkOC1ouTYgL0KsBSSeQAimgsHyYDg2rwCSEploQMJyDCKjQBG5ynPQILwFiHaQAFgISpgiarNAppyA8fw4AUAgdvAyDgOgtzfFi2jfGdshbKapyYOcEb-YCqASlUsAgi8LwSrAZabpcPBYjWP6rEaVBbDYAibrAshlpV3xWMwVgwqypqmqswlGhmHxLCNJzWnNhA8jglwfBm9DoJE5LahG9A4FYAgyHM22mvAMJzC4qyAmWv4DVC5DIIwVDoN8ELM2WHxAsgkQFAAnI24BLGWDwBMbhB1jcHxghmOA8CCGZbB8LjwACkDILADznMQ+7kFsyAgiCUaRMH3tzOQPasl46bIC7JJvmsBQKlsAgvBYPL0FiGaoMgxB-Op5A7KaWIgkI3wPGWVDIJAxA8NamWyAI-ra1QAQLKQMKoH8CzPC8cs4MQENFVdo8ezwAgBBG85UKy0AgiWiCoKyAiAk8KK1pcNjWrNVjVPZjcCN8FenEsNjoIwLzIB8EYsNACy-Ojm7gKaAgFOfZAPJYB3B5JENAkAjjoFQLIHkRpvislWDyciEZIg8meJEQgNxyAaEwP-OYPIXCEGQDwKEGYbjnFZHIDQ+JZCnB2DdeACp7ClHKIEYIoRwhRBiKQAAPPYUAoAljcNiKAFkqgsD0FAAAbx8GUFQ3BFBVC8KAAAvgAGgEUIpY8ixHBBiJgKRwjoixAANoACJWEqHMQAXXsAAPlAJ0IRIi4jiIMVI6A6BYBaNaNo3R7jJEuJMaQCxVjbFaMEX44xPDImgG4OgFQ0REBaISUklQiAADcjhnBLGSFImCkBICgEQCoUgqBEDoFANAORbCYK+FABwdQSR-DIBqOwDUxSBAoHKIgUKFQWG6IzO0jg0BIAPBULAegvCYmiMCYYmRtSFGgCUZ4bwqjHHOOkVo0x4AekZNKKATwoBWSTPHMEnhNiyytG4FkFQU9EnYHmUY1xuz9l9NgHYwRUSIDvNKHEtJyTsmqNMac2A5zZmkBsdkpwYAIyaGiEc7ukAVCyB4coDpjScDVKqSob4IpIBSKsYMthAR8VjOmUadRoBzibNAKYqlNKbF6IkQs0x5xqVGmZX4kIqB-DcBwBSlQMLcn5NAIU4ppTymVNwBS2ARzsWiVSCgNEswcViJCeK8QJS8WoGIKU7A3S2kHOIBUUAVAKiqRUNgXw-hiVWNABmdA6hhmhVGZAGZriWUeMWVYxRIQ1lqLpWSgl0ytHgGIIwVknqQnUssfI8x9jNGCIzLIMgMaeHUtdR0sZEypkZtiPY+xDiRVgDySkFIXjql7hYKobAyTTBVPOTU+gBjcAYHCJinACQWmgDbugU5hr0hIDQFgewsxGAwFKe2x+bqqmtqwDMnRbDnmgFNLAVx9AwkJpsfYgAFFoyQR7oCIHgMIR1zrYDZvdbwoIIQwgRBCbw9dm7qVLGXSoIt3qgm8pUHEvxpi-X+LYQAH3YferhT6X0hPoG+j99ivnfO4IB+R3A72cMfTw59G6YNwfkQh+wABKbg4DVIwoddBnhUjOhaLA7IoD5iIwcDLG0udYzzGZNoPobgz1VBqNo761DoBzHXs6RxrjqgbmgF4-4VRAn6NCfMaaUs1rxOYH0FJmTnH4zLOgbIC1BA5OCLo0s7gInL29vCWo0toAmMFHFf4TEVSWB2r8CsttOAO1zpxdgeAZSVlyH8FampxBsVkAAORSPoNAHAdrkC0Hi-QZA9gF2YD3SJkZYnCOcecBmB4KWDHpadc021SzzHZdALl-LqX0vKe0KpirzgAiIBQIgKLnhGCuYySoSLoB0AcEaZlsZyrjV9NNfQArWB0tMfQCxobkBys5bAM11r7XH5ddKb1-rTb5sjd6f0ibNWFPlE4Ix5jrGc0Lepep1QnBEn2Zk3uwjajGtgDy0AA).

**Comment by Andarist** [comment, marked as answer]
We have some types in the codebase that try to achieve this - I didn't recheck if they are 100% correct but they are probably at least "OK" for the majority of use cases. The problem is that such types, while "cute", can complicate some things:
[TS playground](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgIILAeygJQjAHgBUBRANwnGQgA9IQATAZ2QG9kwBPABwgC5kTMFFABzZAF8AfGwBQyQZQYCAFBDIDSFcAEpkAXhllMwBgG5ZE2bIYQEAGzhQUDuExYBJcNG7PIUYnJKMGo6JRZ2Ll4BIREQcWl5ZGAAW257CBTglnQsXHxA7TAZViSmJVV1TSCqAB9kLWCAbQAiKIgWgF09Q2RjUwsrWXbkAHEIMBqwYhl9BtD6ZjQMbDxCUBhoBqmZAH5touQBEHVoC1kETBAhRSgyYCQDZBOAd2Qvf18J6AJInn5kC0AGIAeRBLUkUhUOgstG42BCIzwLDm40mRQI7UwMFu9yQUgsAHpCQpSWTkAA9XayIA)
[TS playground](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgIILAeygJQjAHgBUBRANwnGQgA9IQATAZ2QG9kwBPABwgC5kTMFFABzZAF8AfGwBQyQZQYCAFBDIDSFcAEpkAXhllMwBgG5ZE2bK69kJOlERgAysAC23ADYRi5SmDUdEos7Lb8gsJikjL6yFoBQfTMyHAgnPLIAPzIANYQnJgw8f5UtMksAEThlZkKOQngdcgCIOrQma3tUBayDBAIXnBQKINwTCwAkuDQ3COQUH7ageUhbBw8EUIiIOLSmR7eEO4BLOhYuPhLATKsmUxKquqapYEAPvaOzm6ePtfgUgA2tVNpUALp6QzIYymCxWGybZAAcQgYFexFi8SSa3O2DwhFAMGgJWWMgarxayDaFB61gQmBAQkUUDIwCQBipEAA7shpgs5qjoAQwpsBJUAGIAeUllRiKh0Flo3GwgXCyDwLDiKLRywI4SKzNZSCkFgA9KaFJarcgAHpZWRAA)

The 2-parameters variant would also suffer from similar problems and it conflicts with our `send` action creator where the second argument configures the send action and not the sent event.

So considering all the things - I kinda think that the ROI of this is rather low here and that the simplicity should win. Always using the explicit object with the `type` property also "optimizes for change" as adding a new property to the event is just adding a new property and not changing a string to an object etc.

**Reply by mellson** [reply]
Thank you for getting back to me.
I understand your position, and I agree that simplicity should probably win here.

I'm not sure which types already help here. The existing type, I could find, for an event to be sent is [this one](https://github.com/statelyai/xstate/blob/main/packages/core/src/types.ts#L81):
```ts
export type Event<TEvent extends EventObject> = TEvent['type'] | TEvent;
```
And if I understand that type correctly, it only checks that the event is an event or has the type property.
But maybe you are thinking about something else?

Anyway, it was a fun little experiment and it shows how far, and crazy, you can get with TypeScript 🎉 

I'll stick to using the entire event object and throw my vote for removing the short-form version entirely for simplicity's sake.

**Reply by Andarist** [reply]
> And if I understand that type correctly, it only checks that the event is an event or has the type property.

Ye, this particular type is unsafe and it will definitely be removed in the future. It's only kept for backward compatibility reasons at the moment.

**Reply by mellson** [reply]
Thanks, great to hear that this is on it's way out 👍🏻
